### PR TITLE
Automated cherry pick of #51921

### LINKED ISSUE #51912: APIServices auto-register controller duels on HA apiserver upgrades
Working through HA upgrade scenarios from 1.7 -> 1.8, I found several situations the current APIService auto-register controller does not handle well:

Health vs liveness:
* Problem: The apiserver health check that waits until local APIServices are registered and available should not go unhealthy if those APIServices are later removed.
* Solution: Make the health check a no-op once it has completed successfully once

CRD flutter:
* Problem: The auto-register controller should not start managing APIService objects until it has completed its initial listing of registered CRD objects and determined the APIService objects it expects to exist.
* Solution: Wait to start the autoregister controller until the CRD controller has completed an initial sync of the initial list of CRD objects.

Controller dueling:
* Problems:
    * If an API server registers APIServices for additional API groups on start, an already-running API server's controller will delete those APIServices.
    * If an API server removes APIServices for API groups no longer served, an already-running API server's controller will re-add those APIServices.
* Solution: expand the controller's automanaged/not-automanaged options to add an "auto-managed on start" option - the controller would sync services with this option exactly once. That allows it to establish services at start, but not duel with API servers started later. In an HA API server setup, that would mean the last API server to start would determine the configured local API services.

The current behavior of the controller:

|                                | Desired: not found          |  Desired: sync always       |
|--------------------------------------|--------------------|--------------------|
|  **Current: error**                               | error              | error              |
|  **Current: not found**                           | noop                | create             |
|  **Current: no sync**                             | noop                | noop                |
|  **Current: sync always**                         | delete             | update             |

The behavior of the controller, adding a "sync on start" option (with the differences in behavior bolded):

|                               | Desired: not found          |  Desired: sync on start     |  Desired: sync always       |
|--------------------------------------|--------------------|--------------------|--------------------|
|  **Current: error**                               | error              | error              | error              |
|  **Current: not found**                           | noop                | **create once** | create             |
|  **Current: no sync**                             | noop                | noop                | noop                |
|  **Current: sync on start, present at start**     | **delete once** | **update once** | **update once** |
|  **Current: sync on start, not present at start** | **noop**                | **noop**                | **noop**                |
|  **Current: sync always**                         | delete             | **update once** | update             |

cc @deads2k @kubernetes/sig-api-machinery-bugs

### PR DESCRIPTION
**Pull request by liggitt**
Cherry pick of https://github.com/kubernetes/kubernetes/pull/51921 on release-1.7.

Fixes #51912

Picking to 1.7.x due to impact on HA upgrades.

/assign @deads2k

cc @kubernetes/sig-api-machinery-bugs

```release-note
Fixes an issue with APIService auto-registration affecting rolling HA apiserver restarts that add or remove API groups being served.
```

### CODE REVIEW

File: pkg/master/thirdparty/tprregistration_controller.go [hunk 1]
```diff
@@ -145,10 +148,39 @@ func (c *tprRegistrationController) Run(threadiness int, stopCh <-chan struct{})
 	defer glog.Infof("Shutting down tpr-autoregister controller")
 
 	// wait for your secondary caches to fill before starting your work
-	if !controller.WaitForCacheSync("tpr-autoregister", stopCh, c.tprSynced) {
+	if !controller.WaitForCacheSync("tpr-autoregister", stopCh, c.tprSynced, c.crdSynced) {
 		return
 	}
 
+	// process each tpr in the list once
```
Comment (liggitt) [inline pkg/master/thirdparty/tprregistration_controller.go]: this was only in the 1.7 version

File: pkg/master/thirdparty/tprregistration_controller.go [hunk 2]
```diff
@@ -145,10 +148,39 @@ func (c *tprRegistrationController) Run(threadiness int, stopCh <-chan struct{})
 	defer glog.Infof("Shutting down tpr-autoregister controller")
 
 	// wait for your secondary caches to fill before starting your work
-	if !controller.WaitForCacheSync("tpr-autoregister", stopCh, c.tprSynced) {
+	if !controller.WaitForCacheSync("tpr-autoregister", stopCh, c.tprSynced, c.crdSynced) {
```
Comment (liggitt) [inline pkg/master/thirdparty/tprregistration_controller.go]: waiting for crdSynced was missing from the 1.7 controller previously

**Comment by liggitt** [comment]
@deads2k PTAL, especially at 1.7-only TPR aspect
@wojtek-t for release approval

**Comment by wojtek-t** [comment]
Cherrypick approved.

@deads2k - can you please LGTM?

**Pull request merged**